src/main/java/com/zaxxer/hikari/pool/PoolBase.java

```diff
@@ -165,8 +165,13 @@ boolean isConnectionAlive(final Connection connection)
             }
          }
          finally {
-            setNetworkTimeout(connection, networkTimeout);
-
+            try {
+               // Some drivers will throw here if the connection is already closed
+               setNetworkTimeout(connection, networkTimeout);
+            }
+            catch (SQLException sqlEx) {
+               return false;
+            }
```

Comment:
The catch for the outer try block is already going to catch this exception and return false. Is the purpose to avoid setting the lastConnectionFailure?

I don't actually understand how this change fixes the issue from #1212 ? Can someone explain it to me?

And it's generally a code smell to return from within a finally block, so my instinct (and likely @brettwooldridge 's as well) is to just say no to any code that does so.

Reply:
Valid concern!
Let's get back to the drawing board.

Seeing that the Postgres JDBC driver [explicitly checks for closed connections](https://github.com/pgjdbc/pgjdbc/blob/master/pgjdbc/src/main/java/org/postgresql/jdbc/PgConnection.java#L1639):
```
Caused by: org.postgresql.util.PSQLException: This connection has been closed.
    at org.postgresql.jdbc.PgConnection.*checkClosed*(PgConnection.java:857)
    at org.postgresql.jdbc.PgConnection.setNetworkTimeout(PgConnection.java:1639)
    at com.zaxxer.hikari.pool.PoolBase.setNetworkTimeout(PoolBase.java:552)
    at com.zaxxer.hikari.pool.PoolBase.isConnectionAlive(PoolBase.java:168)
    at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:185)
```

and that we're already aware of that when explicitly closing connections in https://github.com/brettwooldridge/HikariCP/blob/dev/src/main/java/com/zaxxer/hikari/pool/PoolBase.java#L143
> connection.close(); // continue with the close even if setNetworkTimeout() throws

and

https://github.com/brettwooldridge/HikariCP/blob/dev/src/main/java/com/zaxxer/hikari/pool/PoolBase.java#L531


I see two alternatives:
1) modifying `setNetworkTimeout` to catch the `SQLException` and [return a `boolean` instead](https://github.com/brettwooldridge/HikariCP/blob/dev/src/main/java/com/zaxxer/hikari/pool/PoolBase.java#L555).
The return value can safely be ignored in:
* [quietlyCloseConnection](https://github.com/brettwooldridge/HikariCP/blob/dev/src/main/java/com/zaxxer/hikari/pool/PoolBase.java#L137)
* [isConnectionAlive](https://github.com/brettwooldridge/HikariCP/blob/dev/src/main/java/com/zaxxer/hikari/pool/PoolBase.java#L156-L173)

and continue throwing in case of:

* [resetConnectionState](https://github.com/brettwooldridge/HikariCP/blob/dev/src/main/java/com/zaxxer/hikari/pool/PoolBase.java#L234)
* [setupConnection](https://github.com/brettwooldridge/HikariCP/blob/dev/src/main/java/com/zaxxer/hikari/pool/PoolBase.java#L399-L426)

2) add a `try` method that delegates to previous one, turning the exception to a boolean:
```java
/**
 * Set the network timeout, if <code>isUseNetworkTimeout</code> is <code>true</code> and the
 * driver supports it.
 *
 * @param connection the connection to set the network timeout on
 * @param timeoutMs  the number of milliseconds before timeout
 * @return whether the timeout was successfully applied
 */
private boolean trySetNetworkTimeout(final Connection connection, final long timeoutMs) {
    try {
        // Some drivers will throw here if the connection is already closed
        setNetworkTimeout(connection, timeoutMs);
        return true;
    } catch (SQLException e) {
        return false;
    }
}
```
and using it in`quietlyCloseConnection` and `isConnectionAlive`, ignoring the return value.

What do you think @breskeby, @lfbayer, @johnou, @brettwooldridge?
I'll gladly provide a PR - with tests, if applicable.